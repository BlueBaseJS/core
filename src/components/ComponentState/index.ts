import ComponentState from './ComponentState';
export * from './ComponentState';
export default ComponentState;
