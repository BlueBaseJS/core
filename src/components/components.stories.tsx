import * as React from 'react';
// tslint:disable-next-line:no-implicit-dependencies
import { storiesOf } from '@storybook/react';
import BR from '../index';
import ComponentStateButton from '../components/ComponentState/ComponentStateButton';

const stories = storiesOf('Avatar', module);

stories.add('ComponentStateButton', () => {
	return (
    <ComponentStateButton/>);
});
