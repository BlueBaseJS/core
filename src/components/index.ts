export * from './BlueBaseApp';
export * from './BlueBaseContent';
export * from './BlueBaseFilter';
export * from './ComponentState';
export * from './DataObserver';
export * from './DynamicIcon';
export * from './EmptyState';
export * from './ErrorObserver';
export * from './ErrorState';
export * from './Header';
export * from './HeaderBackButton';
export * from './HeaderTitle';
export * from './HomeScreen';
export * from './HoverObserver';
export * from './JsonSchema';
export * from './Link';
export * from './LoadingState';
export * from './Navigation';
export * from './NavigationActions';
export * from './Noop';
export * from './PluginIcon';
export * from './ReactLoadableLoading';
export * from './Redirect';
export * from './StatefulComponent';
export * from './WaitObserver';
