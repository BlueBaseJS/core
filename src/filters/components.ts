// tslint:disable: object-literal-sort-keys
import {
	ActivityIndicator,
	Button,
	Image,
	ImageBackground,
	Picker,
	Slider,
	Switch,
	Text,
	TextInput,
	View,
} from '../native';
import { BlueBase, BootOptions } from '../BlueBase';
import {
	BlueBaseContent,
	BlueBaseFilter,
	BlueBaseImage,
	ComponentState,
	DataObserver,
	DynamicIcon,
	EmptyState,
	ErrorObserver,
	ErrorState,
	FormattedMessage,
	Header,
	HeaderBackButton,
	HeaderTitle,
	HomeScreen,
	HoverObserver,
	Link,
	LoadingState,
	Navigation,
	NavigationActions,
	Noop,
	PluginIcon,
	Redirect,
	StatefulComponent,
	TouchableItem,
	WaitObserver,
} from '../components/';
import { FilterNestedCollection } from '../registries';
import { ThemeValue } from '../themes';

export const components: FilterNestedCollection = {
	/**
	 * Registers all components that ship with BlueBase
	 */
	'bluebase.components.register.internal': [
		{
			key: 'bluebase-components-register-internal-default',
			priority: 5,

			value: async (bootOptions: BootOptions, _ctx: {}, BB: BlueBase) => {
				await BB.Components.registerCollection({
					// BlueBase Components
					BlueBaseContent,
					BlueBaseFilter,
					BlueBaseImage,
					ComponentState,
					DataObserver,
					DynamicIcon,
					EmptyState,
					ErrorObserver,
					ErrorState,
					FormattedMessage,
					Header,
					HeaderBackButton,
					HeaderTitle,
					HomeScreen,
					HoverObserver,
					Icon: Noop,
					Link,
					LoadingState,
					Navigation,
					NavigationActions,
					Noop,
					PluginIcon,
					Redirect,
					StatefulComponent,
					TouchableItem,
					WaitObserver,

					// Native
					ActivityIndicator,
					Button,
					Image,
					ImageBackground,
					Picker,
					Slider,
					Switch,
					Text,
					TextInput,
					View,

					// Typography
					H1: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.h1 }),
					},
					H2: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.h2 }),
					},
					H3: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.h3 }),
					},
					H4: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.h4 }),
					},
					H5: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.h5 }),
					},
					H6: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.h6 }),
					},
					Subtitle1: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.subtitle1 }),
					},
					Subtitle2: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.subtitle2 }),
					},
					Body1: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.body1 }),
					},
					Body2: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.body2 }),
					},
					Caption: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.caption }),
					},
					Overline: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.overline }),
					},
				});

				return bootOptions;
			},
		},
	],

	/**
	 * This filter registers components from bootOptions.components property.
	 * These are the components typically set in the bluebase.js file.
	 */
	'bluebase.components.register': [
		{
			key: 'bluebase-components-register-default',
			priority: 5,

			// tslint:disable-next-line:object-literal-sort-keys
			value: async (bootOptions: BootOptions, _ctx: {}, BB: BlueBase) => {
				await BB.Components.registerCollection(bootOptions.components);

				return bootOptions;
			},
		},
	],
};
