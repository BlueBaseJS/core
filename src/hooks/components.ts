import { BlueRain, BootOptions } from '../BlueRain';
import { HookInput } from '../registries';
import {
	BlueRainHook,
	ComponentState,
	DynamicIcon,
	EmptyState,
	ErrorState,
	JsonSchema,
	LoadingState,
	Noop,
	PluginIcon,
	SystemApp,
	SystemContent,
} from '../components';

import { ActivityIndicator, Button, Image, Text, View } from '../native';

export const components: { [key: string]: HookInput[] } = {

	'bluerain.components.register.internal': [{
		handler: async (bootOptions: BootOptions, _ctx: {}, BR: BlueRain) => {

			// await BR.Components.register('DynamicIcon', import('../components/DynamicIcon'));
			await BR.Components.register('BlueRainHook', BlueRainHook);
			await BR.Components.register('ComponentState', ComponentState);
			await BR.Components.register('DynamicIcon', DynamicIcon);
			await BR.Components.register('EmptyState', EmptyState);
			await BR.Components.register('ErrorState', ErrorState);
			await BR.Components.register('JsonSchema', JsonSchema);
			await BR.Components.register('LoadingState', LoadingState);
			await BR.Components.register('Noop', Noop);
			await BR.Components.register('PluginIcon', PluginIcon);
			await BR.Components.register('SystemApp', SystemApp);
			await BR.Components.register('SystemContent', SystemContent);
			await BR.Components.register('SystemFooter', Noop);
			await BR.Components.register('SystemHeader', Noop);

			await BR.Components.register('ActivityIndicator', ActivityIndicator);
			await BR.Components.register('Button', Button);
			await BR.Components.register('Image', Image);
			await BR.Components.register('Text', Text);
			await BR.Components.register('View', View);

			return bootOptions;
		},
		name: 'bluerain-components-register-internal-default',
		priority: 5,
	}]
};
