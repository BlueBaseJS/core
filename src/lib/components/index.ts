import {
	ErrorStateProps,
	LoadingStateProps,
	StatefulComponentProps
} from '../../components';
import { TextProps, ViewProps } from 'react-native';
import { BlueBase } from '../../BlueBase';
import { BlueBaseContext } from '../../Context';
import { ButtonProps } from '../../ui-interfaces';
import React from 'react';

export function getComponent<T = any>(name: string) {
	return class BlueBaseComponent extends React.PureComponent<T> {

		static contextType = BlueBaseContext;

		render() {
			const BB: BlueBase = this.context;
			const Component = BB.Components.resolve(name);

			return React.createElement(Component, this.props);
		}
	} as React.ComponentType<T>;
}

// BlueBase
export const ErrorState = getComponent<ErrorStateProps>('ErrorState');
export const LoadingState = getComponent<LoadingStateProps>('LoadingState');
export const StatefulComponent = getComponent<StatefulComponentProps>('StatefulComponent');

// Native
export const Button = getComponent<ButtonProps>('Button');
export const Text = getComponent<TextProps>('Text');
export const View = getComponent<ViewProps>('View');

