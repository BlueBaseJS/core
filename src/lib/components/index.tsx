import { ActivityIndicatorProps, ImageProps, TextProps, ViewProps } from '../../native';
import { BlueBaseConsumer } from '../../Context';
import {
	BlueBaseHookProps,
	ComponentStateProps,
	DataObserverProps,
	DynamicIconProps,
	ErrorObserverProps,
	ErrorStateProps,
	HoverObserverProps,
	JsonSchemaProps,
	LoadingStateProps,
	PluginIconProps,
	StatefulComponentProps,
	WaitObserverProps,
} from '../../components';
import { BlueBase } from '../../BlueBase';
import { ButtonProps } from '../../ui-interfaces';
import React from 'react';

export function getComponent<T = any>(name: string) {

	const BlueBaseComponent = (props: T) => (
		<BlueBaseConsumer children={(BB: BlueBase) => {
			const Component = BB.Components.resolve(name);

			return React.createElement(Component, props);
		}} />
	);

	BlueBaseComponent.displayName = name;

	return BlueBaseComponent as React.ComponentType<T>;
}

// BlueBase
export const BlueBaseHook = getComponent<BlueBaseHookProps>('BlueBaseHook');
export const ComponentState = getComponent<ComponentStateProps>('ComponentState');
export const DataObserver = getComponent<DataObserverProps>('DataObserver');
export const DynamicIcon = getComponent<DynamicIconProps>('DynamicIcon');
export const EmptyState = getComponent<{}>('EmptyState');
export const ErrorObserver = getComponent<ErrorObserverProps>('ErrorObserver');
export const ErrorState = getComponent<ErrorStateProps>('ErrorState');
export const HoverObserver = getComponent<HoverObserverProps>('HoverObserver');
export const JsonSchema = getComponent<JsonSchemaProps>('JsonSchema');
export const LoadingState = getComponent<LoadingStateProps>('LoadingState');
export const PluginIcon = getComponent<PluginIconProps>('PluginIcon');
export const StatefulComponent = getComponent<StatefulComponentProps>('StatefulComponent');
export const WaitObserver = getComponent<WaitObserverProps>('WaitObserver');

// Native
export const ActivityIndicator = getComponent<ActivityIndicatorProps>('ActivityIndicator');
export const Button = getComponent<ButtonProps>('Button');
export const Image = getComponent<ImageProps>('Image');
export const Text = getComponent<TextProps>('Text');
export const View = getComponent<ViewProps>('View');

// Typography
export const H1 = getComponent<TextProps>('H1');
export const H2 = getComponent<TextProps>('H2');
export const H3 = getComponent<TextProps>('H3');
export const H4 = getComponent<TextProps>('H4');
export const H5 = getComponent<TextProps>('H5');
export const H6 = getComponent<TextProps>('H6');
export const Subtitle1 = getComponent<TextProps>('Subtitle1');
export const Subtitle2 = getComponent<TextProps>('Subtitle2');
export const Body1 = getComponent<TextProps>('Body1');
export const Body2 = getComponent<TextProps>('Body2');
export const Caption = getComponent<TextProps>('Caption');
export const Overline = getComponent<TextProps>('Overline');
