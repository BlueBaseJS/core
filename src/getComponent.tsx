import React from 'react';

import { useBlueBase } from './hooks';

/**
 * Resolves a component from BlueBase context, and returns it.
 * It is possible to pass multiple keys as backup. So if one component
 * is not found, the function will fallback and try to find next component.
 *
 * @param keys
 */
export function getComponent<T = any>(
	...keys: Array<string | React.ComponentType<any>>
): React.ComponentType<T> {
	if (keys.length === 0) {
		throw Error('getComponent method needs at least one key');
	}

	const displayName = keys.join('_');

	const BlueBaseComponent = (props: T) => {
		const BB = useBlueBase();

		// If there is no BlueBase context, throw an Error
		if (!BB) {
			throw Error(
				// eslint-disable-next-line max-len
				`Could not resolve component "${displayName}" in "getComponent" command. Reason: BlueBase context not found.`
			);
		}

		return React.createElement(BB.Components.resolveFromCache<T>(...(keys as any)) as any, props as any);
	};

	BlueBaseComponent.displayName = displayName;
	return BlueBaseComponent;
}
