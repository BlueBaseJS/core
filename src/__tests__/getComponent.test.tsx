import { BlueBaseApp } from '../components/';
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { getComponent } from '../getComponent';

describe('getComponent', () => {

	test(`should render a text component`, (done) => {
		const Text = getComponent('Text');

		const component = TestRenderer.create(
			<BlueBaseApp>
				<Text>A Text component</Text>
			</BlueBaseApp>
		);

		setTimeout(() => {
			const tree = component.toJSON();
			expect(tree).toMatchSnapshot();
			expect((tree as any).type).toBe('Text');
			expect((tree as any).children.join()).toBe('A Text component');
			done();
		});
	});

	test(`should render a button component`, (done) => {

		const Button = getComponent('Button');

		const component = TestRenderer.create(
			<BlueBaseApp>
				<Button>A Button component</Button>
			</BlueBaseApp>
		);

		setTimeout(() => {
			const tree = component.toJSON();
			expect(tree).toMatchSnapshot();
			expect((tree as any).children[0].type).toBe('View');
			done();
		});
	});

	test(`should throw an Error when no key is passed`, () => {

		let message;

		try {
			getComponent();
		} catch (error) {
			message = error.message;
		}

		expect(message).toBe('getComponent method needs at least one key');
	});

	// test(`should throw an Error when there is not BlueBase context`, () => {

	// 	const Foo = getComponent('Foo');

	// 	let message;
	// 	try {
	// 		TestRenderer.create(<Foo />);
	// 	} catch (error) {
	// 		message = error.message;
	// 	}

	// 	expect(message)
	// 	.toBe('Could not resolve component in "getComponent" command. Reason: BlueBase context not found.');
	// });

});
