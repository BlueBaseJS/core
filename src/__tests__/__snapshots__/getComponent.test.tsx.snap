// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getComponent should render a button component 1`] = `
<View
  accessibilityRole="button"
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignSelf": "flex-start",
          "borderRadius": 4,
          "paddingBottom": 8,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 8,
        },
        Object {
          "backgroundColor": "#fafafa",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontFamily": "Helvetica",
            "fontSize": 14,
            "fontWeight": "500",
            "textAlign": "center",
            "textTransform": "uppercase",
          },
          Object {},
        ]
      }
    >
      A Button component
    </Text>
  </View>
</View>
`;

exports[`getComponent should render a text component 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "rgba(0, 0, 0, 0.87)",
        "fontFamily": "Helvetica",
        "fontSize": 16,
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  A Text component
</Text>
`;
